### PR TITLE
Add support for SPIR-V extension: SPV_INTEL_function_pointers

### PR DESCRIPTION
**Pull request by VyacheslavLevytskyy**
This PR adds initial support for "SPV_INTEL_function_pointers" SPIR-V extension: https://github.com/intel/llvm/blob/sycl/sycl/doc/design/spirv-extensions/SPV_INTEL_function_pointers.asciidoc

The goal of the extension is to support indirect function calls and translation of function pointers into SPIR-V.

### CODE REVIEW

**Comment by VyacheslavLevytskyy** [comment]
@michalpaszkowski I simplified implementation by removing the feature of resolving opaque pointer types by information about indirect calls. I think it's better to address this later, as just one of many cases when we'd benefit from a more general scheme of improvement of resolving opaque pointer types. This subject appears time from time and probably deserves a generalized solution.

File: llvm/lib/Target/SPIRV/SPIRVCallLowering.cpp [hunk 1]
```diff
@@ -343,9 +348,56 @@ bool SPIRVCallLowering::lowerFormalArguments(MachineIRBuilder &MIRBuilder,
...
+    }
+    // SPIR-V function type:
+    FunctionType *FTy =
+        FunctionType::get(const_cast<Type *>(IC.RetTy), IC.ArgTys, false);
+    SPIRVType *SpirvFuncTy = GR->getOrCreateOpTypeFunctionWithArgs(
+        FTy, SpirvRetTy, SpirvArgTypes, MIRBuilder);
+    // SPIR-V pointer to function type:
+    SPIRVType *IndirectFuncPtrTy = GR->getOrCreateSPIRVPointerType(
+        SpirvFuncTy, MIRBuilder, SPIRV::StorageClass::Function);
+    // Correct the Calee type
```
Comment (michalpaszkowski) [inline llvm/lib/Target/SPIRV/SPIRVCallLowering.cpp:396]: Typo: Calee -> Callee

**Review by michalpaszkowski — APPROVED** [review]
@VyacheslavLevytskyy Thank you for the patch! LGTM! Verified -- the patch does not influence OpenCL CTS pass rate.

File: llvm/lib/Target/SPIRV/SPIRVCallLowering.cpp (hunk 1 above)
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVCallLowering.cpp:396]: Thank you! I'll fix it in the next PR/commit

**Pull request merged**